docs/how-tos/upload-module.md

```diff
@@ -0,0 +1,276 @@
+
+You can choose to upload it to the Viam registry as a _public module_ that is shared with other Viam users, or as a _private module_ that is shared only within your [organization](/cloud/organizations/).
+
+To upload your custom module to the [Viam registry](https://app.viam.com/registry), either as a public or private module, use the Viam CLI commands `create` and `upload` following these instructions:
+
+{{< table >}}
+{{% tablestep %}}
+**1. Install the CLI**
+
+First, [install the Viam CLI](/cli/#install) and [authenticate](/cli/#authenticate) to Viam, from the same machine that you intend to upload your module from.
+
+{{% /tablestep %}}
+{{% tablestep %}}
+**2. Generate a metadata file**
+
+Next, run the `viam module create` command to choose a custom module name and generate the required metadata for your module.
+By default, a new module is created as _private_, meaning that it is only accessible to members of your [organization](/cloud/organizations/), but you can choose to set the `visibility` of your module to _public_ to make it accessible to all Viam users.
+
+Select the private or public tab for instructions to upload your module with the respective `visibility` setting:
+
+{{< tabs >}}
+{{% tab name="Private" %}}
+
+Get the `org-id` for your {{< glossary_tooltip term_id="organization" text="organization" >}} from your organization's **Settings** page in [the Viam App](https://app.viam.com/) and run the following command from the same directory as your custom module to generate metadata for your module:
+
+```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
+viam module create --name <your-module-name> --org-id <your-org-id>
+```
+
+If you later wish to make your module public, you can use the [`viam module update` command](/cli/#module).
+
+{{% /tab %}}
+{{% tab name="Public" %}}
+
+1.  If you haven't already, [create a new namespace](/cloud/organizations/#create-a-namespace-for-your-organization) for your organization.
+    If you have already created a namespace, you can find it on your organization's **Settings** page in [the Viam app](https://app.viam.com/), or by running the [`viam organizations list`](/cli/#organizations) command.
+
+2.  To generate metadata for your module using your public namespace, run the following command from the same directory as your custom module:
+
+    ```sh {id="terminal-prompt" class="command-line" data-prompt="$"}
+    viam module create --name <your-module-name> --public-namespace <your-unique-namespace>
+    ```
+
+{{% /tab %}}
+{{< /tabs >}}
+
+This command creates a new `meta.json` metadata file in your current working directory, which serves as a template.
+
+{{% /tablestep %}}
+{{% tablestep %}}
+**3. Edit the metadata file**
+
+Edit the newly-created `meta.json` file, and provide the required configuration information for your custom module by filling in the following fields.
+
+<table class="table table-striped">
```

Comment:
this is fine for now but we probably want this table to live in reference docs in the future. just pointing it out to ensure we're on the same page